src/ray/core_worker/future_resolver.h

```diff
@@ -0,0 +1,68 @@
+#ifndef RAY_CORE_WORKER_FUTURE_RESOLVER_H
+#define RAY_CORE_WORKER_FUTURE_RESOLVER_H
+
+#include <memory>
+
+#include "ray/common/id.h"
+#include "ray/core_worker/store_provider/memory_store/memory_store.h"
+#include "ray/protobuf/core_worker.pb.h"
+#include "ray/rpc/worker/core_worker_client.h"
+
+namespace ray {
+
+/// Max time between requests to the owner to check whether the object has been
+/// evicted.
+const int kWaitObjectEvictionMilliseconds = 100;
+
+// Resolve values for futures that were given to us before the value
+// was available. This class is thread-safe.
+class FutureResolver {
```

Comment:
The naming is not quite accurate - how about `BorrowedObjectReferenceResolver` or something like that?

Reply:
Hmm I would like to keep the name future to differentiate it from a normal reference (that term is a bit overloaded currently).